luarules/gadgets/unit_weapon_smart_select_helper.lua

```diff
@@ -0,0 +1,216 @@
+function gadget:GetInfo()
+	return {
+		name = "Weapon Smart Select",
+		desc = "Integrates with smart_weapon_select.h and animation scripts to switch firing modes.",
+		author = "SethDGamre",
+		date = "2024.12.7",
+		license = "GNU GPL, v2 or later",
+		layer = 2,
+		enabled = true
+	}
+end
+
+--increment this counter for every hour of your life wasted on smart select: 52
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+--[[
+Integration Checklist:
+1. use weapondef.customparams.smart_preferred_weapon = true for the higher priority smart select weapon.
+2. use weapondef.customparams.smart_deferred_weapon = true for the fallback weapon when smart_preferred_weapon doesn't have a target, or cannot shoot the manually selected target.
+3. use weapondef.customparams.smart_trajectory_checker = true for the weapon that should be used for trajectory checks for the preferredWeapon. Ideally this is a static point slightly lower than preferred_weapon
+3. in the unit's .bos animation script, #include "smart_weapon_select.h"  ideally at the beginning of the file.
+4. in the preferred AimWeaponX() function, add the following at the beginning:
+	if (AimingState != AIMING_PREFERRED){
+		return(0);
+	}
+5. in the deferred AimWeaponX() function, add the following at the beginning:
+	if (AimingState != AIMING_DEFERRED){
+		return(0);
+	}
+6. If using a dummy weapon, return (0); in its AimWeaponX() function and QueryWeaponX(piecenum) should be set to static piece lower than the turret.
+
+****OPTIONAL*****
+use weapondef.customparams.smart_error_frames to override the default reloadtime derivitive frames error threshold
+]]
+
+--static
+local frameCheckModulo = Game.gameSpeed
+local failedToFireMultiplier = Game.gameSpeed * 1.25
+local minimumFailedToFireFrames = Game.gameSpeed * 4
+local aggroDecayRate = 0.65
+local aggroDecayCap = 10
+local tallyDecayRate = 0.99
+local pAutoAggro = 12
+local pManualAggro = pAutoAggro * 4
+local dAutoAggro = 4
+local dManualAggro = pAutoAggro * 3
+local dErrorAggro = -300
+local errorMultiplierAddition = 1
+local switchDeferredThreshold = -dAutoAggro * 1.5
+local switchPreferredThreshold = -1
```

Comment:
These knobs lack documentation so their values are all a bit magic. Describe what the value represents (e.g. "value X is halved each this many seconds") and what the tradeoff is when tweaking the value (e.g. "increase to get better perf, decrease to reduce the risk of inaccurate Y calculation")

Reply:
Also, what do `p` and `d` stand for? Use whole words.

Reply:
Preferred, and deferred. I find it more difficult to read code when the variable names are extra wordy because they also are required to be descriptive. 

Is there a way to get the best of both worlds here?

Reply:
The preferred weapon is tried. If it fails, then it switches to the deferred weapon.
This doesn't have to be just for high/low, it can be for any weapons arrangement (even multiple weapons platforms as long as there's two sets)

Reply:
"Fallback" would be a much better word than "deferred." "Preferred" could be called "primary."

Reply:
Unreadable variable names are bad. Long variable names are bad.

It's a tradeoff, but err on the side of readability, as other devs will need to be able to pick up and work on this code in the future.

Reply:
dictionary definitions
preferred:
like (one thing or person) better than another or others; tend to choose:

deferred:
submit humbly to (a person or a person's wishes or qualities):

shorter and more descriptive than your suggestions. "Primary" implies the first weapon which muddles things. "fallback" is less cool.

All that being said, I think the code is more readable with "p" and "d" prefixes. Perhaps a comment with "p means preferred" before a bunch of variables using "p" would be sufficient to keep code concise.

Acceptable?

Reply:
No. Use the full words, even if they're too long. Using single letters ends up wasting time, not saving it.

Reply:
used "priorityXxxxXxxx" and "backupXxxxxXxxxx" for all named variables. Complete refactor. Tested, seems refactor was successful. Recompilation of COB not necessary because it converts variable names to numbers at runtime anyways.